gax/pagination.h

```diff
@@ -16,60 +16,145 @@
+      using reference = ElementType&;
+
+      ElementType& operator*() const { return *current_; }
+      ElementType* operator->() const { return &(*current_); }
+      iterator& operator++() {
+        ++current_;
+        return *this;
+      }
+      bool operator==(iterator const& rhs) const {
+        return current_ == rhs.current_;
+      }
+      bool operator!=(iterator const& rhs) const { return !(*this == rhs); }
+
+     private:
+      friend PageResult;
+      iterator(FieldIterator current) : current_(std::move(current)) {}
+
+      FieldIterator current_;
+    };
+
+    PageResult(PageType const& raw_page) : raw_page_(raw_page) {}
+    PageResult(PageType&& raw_page) : raw_page_(std::move(raw_page)) {}
+
+    iterator begin() { return ElementAccessor{}(raw_page_)->begin(); }
+    iterator end() { return ElementAccessor{}(raw_page_)->end(); }
+
+    // Const overloads
+    iterator begin() const { return ElementAccessor{}(raw_page_)->cbegin(); }
+    iterator end() const { return ElementAccessor{}(raw_page_)->cend(); }
+
+    std::string NextPageToken() const { return raw_page_.next_page_token(); }
+    PageType const& RawPage() const { return raw_page_; }
+
+   private:
+    friend Pages::iterator;
+    PageType raw_page_;
+  };
+
   class iterator {
    public:
     using iterator_category = std::forward_iterator_tag;
-    using value_type = ElementType;
+    using value_type = PageResult;
     using difference_type = std::ptrdiff_t;
-    using pointer = ElementType*;
-    using reference = ElementType&;
+    using pointer = PageResult*;
+    using reference = PageResult&;
 
-    reference operator*() const { return *current_; }
-    pointer operator->() const { return &(*current_); }
+    PageResult const& operator*() const { return page_result_; }
+    PageResult const* operator->() const { return &page_result_; }
     iterator& operator++() {
-      ++current_;
+      // Note: if the rpc fails, the page will be untouched,
+      // i.e. will be have an empty page token and element collection.
+      // This invalidates any iterators on the PageResult.
+      page_result_.raw_page_.Clear();
+      get_next_page_(&page_result_.raw_page_);
+      num_pages_++;
       return *this;
     }
+
+    // Just want to compare against end()
     bool operator==(iterator const& rhs) const {
-      return current_ == rhs.current_;
+      return page_result_.NextPageToken() == rhs.page_result_.NextPageToken() ||
+             num_pages_ == rhs.num_pages_;
     }
     bool operator!=(iterator const& rhs) const { return !(*this == rhs); }
 
-    iterator(FieldIterator current) : current_(std::move(current)) {}
+   private:
+    friend Pages;
+    // Note: don't need to provide a const-lvalue receiving version of the
+    // constructor since iterators always take ownership of a page and pages can
+    // be very expensive to copy if they have a lot of repeated elements.
+    iterator(PageType&& page_result, NextPageRetriever get_next_page,
```

Comment:
Consider `PageType page_result` and make the caller `std::move()`.

Reply:
Just for my own peace of mind I'll do a quick benchmark first to see if the extra move construction has a noticeable impact. Will update with the results afterwards.

Reply:
In the benchmark, for a repeated field with ~50 elements, having an rvalue ref parameter is about 11% faster. With 100 elements the speedup shrinks to ~1.7%, and with 200 elements it shrinks to less than a percent. Going with the value parameter.